File: apps/admin/types/position.ts
// =====================================================
// 职位数据类型定义
// =====================================================

/**
 * 职位完整类型
 */
export interface Position {
  id: number;
  announcement_id?: number;
  fenbi_announcement_id?: number;
  position_id: string;
  position_code?: string;
  position_name: string;

  // 招录单位
  department_code?: string;
  department_name: string;
  department_level?: string;

  // 招录条件
  recruit_count: number;
  education?: string;
  degree?: string;
  major_category?: string;
  major_requirement?: string;
  major_list?: string[];
  is_unlimited_major: boolean;

  // 工作地点
  work_location?: string;
  province?: string;
  city?: string;
  district?: string;

  // 其他条件
  political_status?: string;
  age?: string;
  age_min: number;
  age_max: number;
  work_experience?: string;
  work_experience_years: number;
  is_for_fresh_graduate?: boolean | null;
  gender?: string;
  household_requirement?: string;
  service_period?: string;
  other_conditions?: string;

  // 考试信息
  exam_type?: string;
  exam_category?: string;

  // 时间信息
  registration_start?: string;
  registration_end?: string;
  exam_date?: string;
  interview_date?: string;

  // 其他信息
  salary_range?: string;
  remark?: string;
  source_url?: string;

  // 报名统计
  applicant_count: number;
  pass_count: number;
  competition_ratio: number;

  // AI解析元数据
  parse_confidence: number;
  parsed_at?: string;

  // 状态
  status: number;

  // 系统字段
  created_at: string;
  updated_at: string;
}

/**
 * 职位简要类型（列表用）
 */
export interface PositionBrief {
  id: number;
  position_name: string;
  position_code?: string;
  department_name: string;
  department_level?: string;
  recruit_count: number;
  education?: string;
  is_unlimited_major: boolean;
  province?: string;
  city?: string;
  exam_type?: string;
  registration_end?: string;
  exam_date?: string;
  competition_ratio: number;
  is_for_fresh_graduate?: boolean | null;
  status: number;
  created_at: string;
}

/**
 * 职位详情类型
 */
export interface PositionDetail extends Position {
  is_favorite?: boolean;
  announcements?: Array<{
    id: number;
    title: string;
    source_url?: string;
    publish_date?: string;
  }>;
}

/**
 * 职位查询参数类型
 */
export interface PositionQueryParams {
  // 地域筛选
  province?: string;
  city?: string;
  district?: string;
  provinces?: string[];

  // 条件筛选
  education?: string;
  educations?: string[];
  degree?: string;
  major?: string;
  major_category?: string;
  major_keyword?: string;
  exam_type?: string;
  exam_types?: string[];
  exam_category?: string;
  department_level?: string;
  political_status?: string;
  gender?: string;

  // 快捷筛选
  unlimited_major?: boolean;
  fresh_graduate?: boolean;
  no_experience?: boolean;
  min_recruit?: number;

  // 时间筛选
  reg_status?: "registering" | "upcoming" | "ended";
  expiring_days?: number;
  updated_today?: boolean;

  // 关键词
  keyword?: string;
  dept_keyword?: string;

  // 分页排序
  page?: number;
  page_size?: number;
  sort_by?: "recruit_count" | "created_at" | "registration_end" | "match_score";
  sort_order?: "asc" | "desc";

  // 状态筛选
  status?: number;
}

/**
 * 职位列表响应类型
 */
export interface PositionListResponse {
  positions: PositionBrief[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * 职位统计数据类型
 */
export interface PositionStats {
  total_count: number;
  published_count: number;
  today_new_count: number;
  registering_count: number;
  expiring_count: number;
  total_recruit_count: number;
  avg_competition: number;
}

/**
 * 按省份统计
 */
export interface ProvinceStats {
  province: string;
  count: number;
  recruit_count: number;
}

/**
 * 按考试类型统计
 */
export interface ExamTypeStats {
  exam_type: string;
  count: number;
  recruit_count: number;
}

/**
 * 按学历统计
 */
export interface EducationStats {
  education: string;
  count: number;
}

/**
 * 趋势数据
 */
export interface TrendItem {
  date: string;
  count: number;
}

/**
 * 趋势响应
 */
export interface PositionTrendsResponse {
  trends: TrendItem[];
}

/**
 * 筛选选项类型
 */
export interface FilterOptions {
  provinces: string[];
  exam_types: string[];
  exam_categories: string[];
  major_categories: string[];
  department_levels: string[];
  educations: string[];
  political_status: string[];
  genders: string[];
}

/**
 * 级联地区类型
 */
export interface CascadeRegion {
  province: string;
  cities: string[];
}

/**
 * 级联地区响应
 */
export interface CascadeRegionsResponse {
  regions: CascadeRegion[];
}

/**
 * 职位状态枚举
 */
export enum PositionStatus {
  Pending = 0,
  Published = 1,
  Offline = 2,
}

/**
 * 职位状态名称映射
 */
export const PositionStatusNames: Record<PositionStatus, string> = {
  [PositionStatus.Pending]: "待审核",
  [PositionStatus.Published]: "已发布",
  [PositionStatus.Offline]: "已下线",
};

/**
 * 职位状态颜色映射
 */
export const PositionStatusColors: Record<PositionStatus, string> = {
  [PositionStatus.Pending]: "bg-amber-100 text-amber-700",
  [PositionStatus.Published]: "bg-green-100 text-green-700",
  [PositionStatus.Offline]: "bg-gray-100 text-gray-700",
};

/**
 * 学历选项
 */
export const EducationOptions = [
  "大专",
  "本科",
  "硕士研究生",
  "博士研究生",
];

/**
 * 学位选项
 */
export const DegreeOptions = [
  "学士",
  "硕士",
  "博士",
];

/**
 * 考试类型选项
 */
export const ExamTypeOptions = [
  "公务员",
  "事业单位",
  "教师招聘",
  "医疗卫生",
  "银行招聘",
  "国企招聘",
  "军队文职",
  "三支一扶",
  "选调生",
  "社区工作者",
];

/**
 * 考试分类选项 (A/B/C类)
 */
export const ExamCategoryOptions = [
  "综合管理类A",
  "社会科学专技类B",
  "自然科学专技类C",
  "中小学教师类D",
  "医疗卫生类E",
];

/**
 * 单位层级选项
 */
export const DepartmentLevelOptions = [
  "中央",
  "省级",
  "市级",
  "县级",
  "乡镇",
];

/**
 * 政治面貌选项
 */
export const PoliticalStatusOptions = [
  "不限",
  "中共党员",
  "中共党员或共青团员",
  "共青团员",
  "群众",
];

File: apps/server/internal/model/position.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// Position 职位模型
type Position struct {
	ID                 uint           `gorm:"primaryKey" json:"id"`
	AnnouncementID     *uint          `gorm:"index" json:"announcement_id"`                                                            // 关联公告ID
	FenbiAnnouncementID *uint         `gorm:"index" json:"fenbi_announcement_id"`                                                      // 粉笔公告ID
	PositionID         string         `gorm:"type:varchar(50);uniqueIndex;not null" json:"position_id"`                                // 职位唯一标识
	PositionCode       string         `gorm:"type:varchar(50);index" json:"position_code"`                                             // 职位代码
	PositionName       string         `gorm:"type:varchar(200);not null;index" json:"position_name" es:"type:text,analyzer:ik_max_word"` // 岗位名称

	// 招录单位
	DepartmentCode  string `gorm:"type:varchar(50);index" json:"department_code"`                                     // 单位代码
	DepartmentName  string `gorm:"type:varchar(200);index" json:"department_name" es:"type:text,analyzer:ik_max_word"` // 招录单位
	DepartmentLevel string `gorm:"type:varchar(50)" json:"department_level"`                                           // 单位层级

	// 招录条件
	RecruitCount     int             `gorm:"default:1" json:"recruit_count"`                   // 招录人数
	Education        string          `gorm:"type:varchar(50);index" json:"education"`         // 学历要求
	Degree           string          `gorm:"type:varchar(50)" json:"degree"`                  // 学位要求
	MajorCategory    string          `gorm:"type:varchar(100)" json:"major_category"`         // 专业大类
	MajorRequirement string          `gorm:"type:text" json:"major_requirement"`              // 专业要求原文
	MajorList        JSONStringArray `gorm:"type:json" json:"major_list"`                     // 专业列表JSON
	IsUnlimitedMajor bool            `gorm:"default:false;index" json:"is_unlimited_major"`   // 是否不限专业

	// 工作地点
	WorkLocation string `gorm:"type:varchar(200)" json:"work_location"` // 工作地点
	Province     string `gorm:"type:varchar(50);index" json:"province"` // 省份
	City         string `gorm:"type:varchar(50);index" json:"city"`     // 城市
	District     string `gorm:"type:varchar(50)" json:"district"`       // 区县

	// 其他条件
	PoliticalStatus      string `gorm:"type:varchar(50)" json:"political_status"`          // 政治面貌要求
	Age                  string `gorm:"type:varchar(100)" json:"age"`                      // 年龄要求原文
	AgeMin               int    `gorm:"default:18" json:"age_min"`                         // 最小年龄
	AgeMax               int    `gorm:"default:35" json:"age_max"`                         // 最大年龄
	WorkExperience       string `gorm:"type:varchar(200)" json:"work_experience"`          // 工作经历要求原文
	WorkExperienceYears  int    `gorm:"default:0" json:"work_experience_years"`            // 最低工作年限
	IsForFreshGraduate   *bool  `gorm:"index" json:"is_for_fresh_graduate"`                // 是否限应届(NULL=不限,true=仅应届,false=需经验)
	Gender               string `gorm:"type:varchar(10)" json:"gender"`                    // 性别要求
	HouseholdRequirement string `gorm:"type:varchar(200)" json:"household_requirement"`    // 户籍要求
	ServicePeriod        string `gorm:"type:varchar(100)" json:"service_period"`           // 服务期限
	OtherConditions      string `gorm:"type:text" json:"other_conditions"`                 // 其他条件

	// 考试信息
	ExamType     string `gorm:"type:varchar(50);index" json:"exam_type"`  // 考试类型
	ExamCategory string `gorm:"type:varchar(50)" json:"exam_category"`    // 考试分类(A/B/C类)

	// 时间信息
	RegistrationStart *time.Time `gorm:"type:datetime;index" json:"registration_start"` // 报名开始时间
	RegistrationEnd   *time.Time `gorm:"type:datetime;index" json:"registration_end"`   // 报名截止时间
	ExamDate          *time.Time `gorm:"type:datetime" json:"exam_date"`                // 笔试时间
	InterviewDate     *time.Time `gorm:"type:datetime" json:"interview_date"`           // 面试时间

	// 其他信息
	SalaryRange string `gorm:"type:varchar(100)" json:"salary_range"` // 薪资范围
	Remark      string `gorm:"type:text" json:"remark"`               // 备注
	SourceURL   string `gorm:"type:varchar(500)" json:"source_url"`   // 来源链接

	// 报名统计(可选)
	ApplicantCount   int     `gorm:"default:0" json:"applicant_count"`                       // 报名人数
	PassCount        int     `gorm:"default:0" json:"pass_count"`                            // 过审人数
	CompetitionRatio float64 `gorm:"type:decimal(10,2);default:0" json:"competition_ratio"`  // 竞争比

	// AI解析元数据
	ParseConfidence int        `gorm:"default:0" json:"parse_confidence"` // 解析置信度(0-100)
	ParsedAt        *time.Time `gorm:"type:datetime" json:"parsed_at"`    // 解析时间

	// 状态
	Status int `gorm:"type:tinyint;default:0;index" json:"status"` // 0: pending, 1: published, 2: offline

	// 系统字段
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	Announcements []Announcement `gorm:"many2many:what_position_announcements;" json:"announcements,omitempty"`
}

func (Position) TableName() string {
	return "what_positions"
}

// PositionBriefResponse 职位简要信息（列表用）
type PositionBriefResponse struct {
	ID                 uint       `json:"id"`
	PositionName       string     `json:"position_name"`
	PositionCode       string     `json:"position_code,omitempty"`
	DepartmentName     string     `json:"department_name"`
	DepartmentLevel    string     `json:"department_level,omitempty"`
	RecruitCount       int        `json:"recruit_count"`
	Education          string     `json:"education,omitempty"`
	IsUnlimitedMajor   bool       `json:"is_unlimited_major"`
	Province           string     `json:"province,omitempty"`
	City               string     `json:"city,omitempty"`
	ExamType           string     `json:"exam_type,omitempty"`
	RegistrationEnd    *time.Time `json:"registration_end,omitempty"`
	CompetitionRatio   float64    `json:"competition_ratio,omitempty"`
	IsForFreshGraduate *bool      `json:"is_for_fresh_graduate,omitempty"`
	Status             int        `json:"status"`
	CreatedAt          time.Time  `json:"created_at"`
}

// ToBriefResponse 转换为简要响应
func (p *Position) ToBriefResponse() *PositionBriefResponse {
	return &PositionBriefResponse{
		ID:                 p.ID,
		PositionName:       p.PositionName,
		PositionCode:       p.PositionCode,
		DepartmentName:     p.DepartmentName,
		DepartmentLevel:    p.DepartmentLevel,
		RecruitCount:       p.RecruitCount,
		Education:          p.Education,
		IsUnlimitedMajor:   p.IsUnlimitedMajor,
		Province:           p.Province,
		City:               p.City,
		ExamType:           p.ExamType,
		RegistrationEnd:    p.RegistrationEnd,
		CompetitionRatio:   p.CompetitionRatio,
		IsForFreshGraduate: p.IsForFreshGraduate,
		Status:             p.Status,
		CreatedAt:          p.CreatedAt,
	}
}

// PositionDetailResponse 职位详情响应
type PositionDetailResponse struct {
	Position
	IsFavorite bool `json:"is_favorite,omitempty"`
}

// ToDetailResponse 转换为详情响应
func (p *Position) ToDetailResponse(isFavorite bool) *PositionDetailResponse {
	return &PositionDetailResponse{
		Position:   *p,
		IsFavorite: isFavorite,
	}
}
